File: LeetCode/Tree/Minimum_Depth_of_Binary_Tree.cpp
/*Given a binary tree, find its minimum depth.

The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
*/

#include <iostream>
#include <stdlib.h>
using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x):val(x),left(NULL),right(NULL){}
};

//using recursive
class Solution1{
public:
    int minDepth(TreeNode *root){
        if(root==NULL){
            return 0;
        }
        if(root->left==NULL){
            return 1+minDepth(root->right);
        }
        if(root->right==NULL){
            return 1+minDepth(root->left);
        }
        return 1+min(minDepth(root->left),minDepth(root->right));
    }
    int min(int a, int b){
        return a<b?a:b;
    }
};

TreeNode *newNode(int data){
    TreeNode *node=(TreeNode*)malloc(sizeof(TreeNode));
    node->val=data;
    node->left=NULL;
    node->right=NULL;
}

int main(){
    Solution1 solution1;
    TreeNode *root=newNode(1);
    root->left=newNode(2);
    root->right=newNode(3);
    root->left->left=newNode(4);
    int min=solution1.minDepth(root);
    cout<<min<<endl; //2
    return 0;
}

File: LeetCode_Python/Tree/Minimum_Depth_of_Binary_Tree.py
'''
Given a binary tree, find its minimum depth.

The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
'''

'''
class TreeNode(object):
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None
'''

class Solution(object):
	#DFS
	def minDepth(self, root):
		if root is None:
			return 0
		left = self.minDepth(root.left)
		right = self.minDepth(root.right)
		if left and right:
			return min(left, right) + 1
		return max(left, right) + 1

	#BFS
	